recording_external_revenue/models/recording_external_revenue_raw.py

```diff
@@ -55,37 +56,163 @@ class RecordingExternalRevenueRaw(models.Model):
     commission_amount = fields.Float("Total Commissions Amount")
     net_amount = fields.Float("Total Net Amount (Untaxed)")
 
-    def make_revenue(self):
+    @api.multi
+    def name_get(self):
+        return [(r.id, "#{}".format(r.id)) for r in self]
+
+    @api.multi
+    def write(self, vals):
+        self._check_can_edit_fields(vals)
+        return super().write(vals)
+
+    @api.multi
+    def unlink(self):
+        self._check_can_unlink()
+        return super().unlink()
+
+    def _check_can_edit_fields(self, vals):
+        fields_to_check = self._get_fields_protected_after_conversion()
+        converted_revenues = self.filtered("is_converted")
+        if converted_revenues and fields_to_check.intersection(vals):
+            raise ValidationError(
+                _(
+                    "The following raw revenue lines can not be edited because "
+                    "these are already converted into revenues:\n"
+                    "{}"
+                ).format(", ".join(converted_revenues.mapped("display_name")))
+            )
+
+    def _get_fields_protected_after_conversion(self):
+        result = set(self.get_direct_mapping_fields())
+        result.update(self.get_aggregated_fields())
+        result.update(self.get_common_parameter_fields())
+        return result
+
+    def _check_can_unlink(self):
+        converted_revenues = self.filtered("is_converted")
+        if converted_revenues:
+            raise ValidationError(
+                _(
+                    "The following raw revenue lines can not be deleted because "
+                    "these are already converted into revenues:\n"
+                    "{}"
+                ).format(", ".join(converted_revenues.mapped("display_name")))
+            )
+
+    def schedule_conversion(self, company):
+        revenues_to_convert = self.search([("is_converted", "=", False)]).filtered(
+            lambda r: r.company_id == company
+        )
+        for revenues in revenues_to_convert.group_by_common_parameters():
+            revenues.with_delay().convert()
+
+    def group_by_common_parameters(self):
+        fields = self.get_common_parameter_fields()
+
+        def groupby_func(record):
+            return tuple(record[f] for f in fields)
+
+        result = {}
+
+        for record in self:
+            key = groupby_func(record)
+            if key in result:
+                result[key] |= record
+            else:
+                result[key] = record
+
+        return result.values()
+
+    def get_common_parameter_fields(self):
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Method could be a function](https://app.codacy.com/manual/numigi/odoo-recording-addons/pullRequest?prid=5620550)